Sources/PackageLoading/PackageBuilder.swift

```diff
@@ -382,7 +389,15 @@ public struct PackageBuilder {
             throw ModuleError.modulesNotFound(missingModuleNames)
         }
 
-        return modules
+        // Normally, test modules are only dependent upon modules with
+        // the same basename. For example, a test module in
+        // 'Root/Tests/FooTests' is dependent upon 'Root/Sources/Foo'.
+        // Only do this if there is no explict dependency declared in manifest.
```

Comment:
Thats a good idea but it'll be a whole new activity. Currently how it works is, first we construct modules, fill the dependencies from manifest if declared and then see which (test) modules are left with zero deps. This entire thing, the PackageDescription module, TOML and maybe some other things will need refactoring.